node/node.go

```diff
@@ -374,6 +383,60 @@ func (n *Node) initNetworking(primaryNetVdrs validators.Set) error {
 	return err
 }
 
+type NodeRuntimeState struct {
+	// The process id of the node
+	PID int
+	// URI to access the node API
+	// Format: [https|http]://[host]:[port]
+	URI string
+	// Address other nodes can use for bootstrapping
+	// Format: [host]:[port]
+	BootstrapAddress string
+}
+
+// Write runtime state to the configured path. Supports the use of
+// dynamically chosen network ports with local network orchestration.
+func (n *Node) writeRuntimeState() {
+	n.Log.Info("attempting to write runtime state", zap.String("path", n.Config.RuntimeStatePath))
+
+	uri := ""
+
+	// Wait until the API Server URI is available.
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
+	defer cancel()
+	for {
+		select {
+		case <-ctx.Done():
+			n.Log.Error("failed to retrieve API Server URI before timeout")
+			return
+		default:
+			uri = n.APIServer.GetURI()
+		}
+		if len(uri) > 0 {
+			break
+		}
+		time.Sleep(time.Millisecond * 200)
+	}
+
+	// Write the runtime state to disk
+	runtimeState := &NodeRuntimeState{
+		PID:              os.Getpid(),
+		URI:              uri,
+		BootstrapAddress: n.bootstrapAddress, // Set by network initialization
+	}
+	bytes, err := json.MarshalIndent(runtimeState, "", "  ")
+	if err != nil {
+		n.Log.Error("failed to marshal runtime state", zap.Error(err))
+		return
+	}
+	if err := os.WriteFile(n.Config.RuntimeStatePath, bytes, perms.ReadWrite); err != nil {
+		n.Log.Error("failed to write runtime state", zap.Error(err))
+		return
+	}
+
+	n.Log.Info("wrote runtime state")
+}
```

Comment:
I know this was done in an attempt to be a minimally invasive as possible... But I think we should be more invasive.

Rather than using `tls.Listen` we can simply move the `net.Listen` into `server.New` so that we can expose the hostname without needing to handle any threading logic.